Services/Statistics/SimulationStatistics.cs

```diff
@@ -91,8 +92,17 @@ public async Task CreateOrUpdateAsync(string simulationId, SimulationStatisticsM
 
             try
             {
-                this.log.Debug("Creating statistics record", () => new { statisticsStorageRecord });
-                await this.simulationStatisticsStorage.CreateAsync(statisticsStorageRecord);
+                if (await this.simulationStatisticsStorage.ExistsAsync(statisticsRecordId))
```

Comment:
I would try to get the existing record, if it exists, and then pass the ETag. If it doesn't exist, you could pass a wildcard for the ETag. The storage library will handle this and will do the right thing with the DocumentDB API.

`var ETag = "*";
var existingRecord = await this.simulationStatisticsStorage.ExistsAsync(statisticsRecordId);
if(existingRecord != null) ETag = existingRecord.ETag;
await this.simulationStatisticsStorage.UpsertAsync(statisticsStorageRecord, ETag);`

Reply:
According to the [DocumentDB documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.requestoptions?view=azure-dotnet), you can send a meaningless ETag for creation; it'll be ignored. 

"ETag, while valid on Replace* and Delete* operations, would have no impact on a Read*, CreateQuery* or Create* operations."

Reply:
Good point about using Upsert for both cases.  In both cases we have similar checks for Exists and pass expected eTag and also it seems cleaner to call Create if the record does not exist. I prefer to keep this implementation.